ql/cashflows/fixedratecoupon.cpp

```diff
@@ -175,30 +175,20 @@ namespace QuantLib {
                                                      exCouponAdjustment_,
                                                      exCouponEndOfMonth_);
         }
-
-        if (schedule_.isRegular(1)) {
-            QL_REQUIRE(firstPeriodDC_.empty() ||
-                       firstPeriodDC_ == rate.dayCounter(),
-                       "regular first coupon "
-                       "does not allow a first-period day count");
-            shared_ptr<CashFlow> temp(new
-                FixedRateCoupon(paymentDate, nominal, rate,
-                                start, end, start, end, exCouponDate));
-            leg.push_back(temp);
-        } else {
-            Date ref = schedule_.calendar().advance(
-                                            end,
-                                            -schedule_.tenor(),
-                                            schedule_.businessDayConvention(),
-                                            schedule_.endOfMonth());
-            InterestRate r(rate.rate(),
-                           firstPeriodDC_.empty() ? rate.dayCounter()
-                                                  : firstPeriodDC_,
-                           rate.compounding(), rate.frequency());
-            leg.push_back(shared_ptr<CashFlow>(new
-                FixedRateCoupon(paymentDate, nominal, r,
-                                start, end, ref, end, exCouponDate)));
-        }
+        Date ref = schedule_.hasTenor() &&
+            !(schedule_.hasIsRegular() && schedule_.isRegular(1)) ?
```

Comment:
I think you mean `hasTenor() && hasIsRegular() && !isRegular(1)`.  As written, if `hasIsRegular()` is false it tries to do the `advance`.

I also think we need a few test cases, at least for the fixed and floating legs...

Reply:
You are right. I hate this..., will change it. I agree with regards to the unit tests, I'll try to add a few over the weekend...